flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java

```diff
@@ -101,26 +140,20 @@ public void snapshotState(FunctionSnapshotContext context) throws Exception {
                 this.checkpointedState != null,
                 "The " + getClass().getSimpleName() + " state has not been properly initialized.");
 
-        this.checkpointedState.update(new ArrayList<>(this.valuesToEmit));
+        this.checkpointedState.clear();
+        this.checkpointedState.addAll(new ArrayList<>(subTaskStates));
```

Comment:
```suggestion
        this.checkpointedState.update(new ArrayList<>(subTaskStates));
```
nit: we can continue to rely on `update` here, I guess, to keep the diff smaller.